src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(currentState, stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(currentState, stateCopy, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(currentState, stateCopy);
+        break;
+      default:
+        break;
```

Comment:
The `default` case in the `switch` statement is empty. It would be beneficial to include error handling here in case an action with an unsupported `type` is encountered. This could be as simple as throwing an error or logging a warning message.